containers/daap-python-base/src/var/task/data_product_metadata.py

```diff
@@ -0,0 +1,117 @@
+
+
+def get_data_product_metadata_spec_path(version: str = None) -> str:
+    # if version is None we'll get the latest version
+    if version is None:
+        file_paths = get_filepaths_from_s3_folder(
+            os.path.join("s3://", get_bucket_name(), "data_product_metadata_spec")
+        )
+        versions = list(
+            {i for p in file_paths for i in p.split("/") if i.startswith("v")}
+        )
+        versions.sort()
+        latest_version = versions[-1]
+        spec_path = os.path.join("s3://", get_bucket_name(), "data_product_metadata_spec", latest_version, "moj_data_product_metadata_spec.json")
+    else:
+        spec_path = os.path.join("s3://", get_bucket_name(), "data_product_metadata_spec", version, "moj_data_product_metadata_spec.json")
+
+    return spec_path
+
+
+def split_bucket_and_key(path):
+    bucket, key = path.replace("s3://", "").split("/", 1)
+    return bucket, key
+
+
+class DataProductMetadata():
+    """
+    class to handle creation and updating of
+    metadata relating to a dataproduct
+    """
+    def __init__(self, data_product_name: str, logger: DataPlatformLogger):
+        self.data_product_name = data_product_name
+        self.logger = logger
+        bucket, key = split_bucket_and_key(get_data_product_metadata_path(data_product_name))
+        self.metadata_bucket = bucket
+        self.metadata_key = key
+        self._check_if_metadata_exists()
+
+
+    def _check_if_metadata_exists(self) -> bool:
+        # establish whether metadata for data product already exists
+        try:
+            # get head of object (if it exists)
+            s3_client.head_object(Bucket=self.metadata_bucket, Key=self.metadata_key)
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                self.metadata_exists = False
+            else:
+                self.logger.error(f"Uknown error - {e}")
+                raise Exception(f"Uknown error - {e}")
+        else:
+            self.logger.info("version 1 of metadata already exists for this data product")
+            self.metadata_exists = True
+
+        return self
+
+    def validate_metadata(
+        self,
+        data_product_metadata: dict,
+        metadata_schema_version: str = None,
+    ) -> bool:
+
+        metadata_schema = read_json_from_s3(get_data_product_metadata_spec_path(metadata_schema_version))
+        try:
+            validate(instance=data_product_metadata, schema=metadata_schema)
+        except ValidationError:
+            self.error_traceback = traceback.format_exc()
+            self.logger.error(f"metadata has failed validation with error: {self.error_traceback}")
+            self.valid_metadata = False
+        else:
+            self.logger.info("metadata has passed validation")
+            self.valid_metadata = True
+            self.data_product_metadata = data_product_metadata
+        return self
+
+    def write_metadata_json_to_s3(self) -> None:
```

Comment:
I think this method has an assumption that the validate method has already been called first and the validation didn't fail.

If that's the case, I think this method should check that `data_product_metadata` is not None first, and raise some kind of validation error if not.